mypy_json_report.py

```diff
@@ -71,59 +134,144 @@ def _no_command(args: argparse.Namespace) -> None:
     """
-    Produce a structure to hold the mypy errors.
+    Produces a summary of errors in a Mypy report.
 
-    The resulting structure looks like this:
+    The structure of grouped_errors looks like this once processing is complete:
 
         {
-            "module/filename.py": [
+            "module/filename.py": {
                 "Mypy error message": 42,
                 "Another error message": 19,
                 ...
-            ],
+            },
             ...
         }
     """
-    grouped_errors: Dict[str, Dict[str, int]] = defaultdict(dict)
-    for error, frequency in errors.items():
-        grouped_errors[error.filename][error.message] = frequency
 
-    return dict(grouped_errors)
+    def __init__(self) -> None:
+        self.grouped_errors: ErrorSummary = defaultdict(Counter)
+
+    def process_message(self, message: MypyMessage) -> None:
+        if message.message_type != "error":
+            return None
+        self.grouped_errors[message.filename][message.message] += 1
+
+
+@dataclass(frozen=True)
+class DiffReport:
+    error_lines: Tuple[str, ...]
+    total_errors: int
+    num_new_errors: int
+    num_fixed_errors: int
+
+
+class ChangeTracker:
+    """
+    Compares the current Mypy report against a previous summary.
+
+    Stores errors as it goes so that raw messages can be shown when a diff is found.
+
+    Relies on the fact that Mypy reports are grouped by file to reduce messages
+    that are cached in memory.
+    """
+
+    def __init__(self, summary: ErrorSummary) -> None:
+        self.old_report = summary
+        self.error_lines: List[str] = []
+        self.current_file: Optional[str] = None
+        self.num_errors = 0
+        self.num_new_errors = 0
+        self.num_fixed_errors = 0
+        self._reset_caches()
+
+    def process_message(self, message: MypyMessage) -> None:
+        """Track a MypyMessage."""
+        if self.current_file != message.filename:
+            self._flush_batch()
+            self.current_file = message.filename
+
+        self._messages_by_line_number[message.line_number].append(message.raw)
+        if message.message_type == "error":
+            self.num_errors += 1
+            self._error_frequencies[message.message] += 1
+            self._line_numbers_by_error[message.message].append(message.line_number)
+
+    def diff_report(self) -> DiffReport:
+        self._flush_batch()
+        return DiffReport(
+            error_lines=tuple(self.error_lines),
+            total_errors=self.num_errors,
+            num_new_errors=self.num_new_errors,
+            num_fixed_errors=self.num_fixed_errors,
+        )
+
+    def _reset_caches(self) -> None:
+        """Reset the per-file caches, ready for the next file."""
+        self._messages_by_line_number: Dict[int, List[str]] = defaultdict(list)
+        self._line_numbers_by_error: Dict[str, List[int]] = defaultdict(list)
+        self._error_frequencies: CounterType[str] = Counter()
```

Comment:
Note: I know I can simplify this by using `itertools.groupby` in `_parse_command`, but I wasn't sure which was better.

TODO: Add that commit to another branch for comparison.

Reply:
I decided I liked it, so it's on this branch now.